Support open amount invoice refund

bhelx opened the pull request:
Resolves #169 

This PR is intended to support both open amount refunds on invoices. It's still a work in progress. I've decided to move the other refund method to a follow up PR.

pierre commented on src/test/java/com/ning/billing/recurly/TestRecurlyClient.java:
```diff
@@ -154,6 +157,7 @@ public void testRemoveSubscriptionAddons() throws Exception {
 
             // Create a plan with addons
             final Plan plan = recurlyClient.createPlan(planData);
+            Thread.sleep(1000);
```
There is no better synchronization option? Could we poll the server instead maybe?

bhelx replied:
Yes, not sure why this started happening all of the sudden but I am investigating why and trying to determine a better solution.

pierre replied:
👍

bhelx replied:
Okay, I figured out this is a temporary thing on the server side. It should be operating normally by next week. I'm going to add a TODO tag here and get rid of this next week.

pierre commented on src/main/java/com/ning/billing/recurly/RecurlyClient.java:
```diff
@@ -773,6 +794,24 @@ public Invoices getAccountInvoices(final String accountCode) {
+     * @param order If credit line items exist on the invoice, this parameter specifies which refund method to use first
+     * @return the refunded invoice
+     */
+    public Invoice refundInvoice(final String invoiceId, final Integer amountInCents, final RefundApplyOrder order) {
```
Can an invoice have several (partial) payments?

bhelx replied:
You can only refund an invoice that was successfully paid. This method allows you to refund an open amount on the invoice. Since the invoice may contain transactions and credits, it allows you to define which to prioritize in the refund. We have another method that allows you to pass in specific transaction refund objects but I've decided to add that one in a follow up.

bhelx commented:
okay @pierre 

Removed WIP and hopefully addressed some of your comments. I'll have the follow up for the other refund method out this week and I can remove those `Thread.sleep()` next week.

pierre commented:
> I'll have the follow up for the other refund method out this week and I can remove those `Thread.sleep()` next week.

👍 Once we have these, I'll merge and trigger a release.

pierre commented:
@bhelx Any update on the `sleep`?

cfrasineanu-intx commented:
Our project will need this change soon. Do you have the timeline? Thank you

bhelx commented:
@pierre I was told that the change needed to remove the sleep was pushed back two weeks. I think it's probably best to leave the sleeps then I can come back with a follow up in two weeks.

pierre commented:
Aww 😿 

Alright, I'll merge it as-is to unblock @cfrasineanu-intx.

pierre commented:
@cfrasineanu-intx 0.8.0 is on its way to Maven Central.

bhelx commented:
Thanks @pierre!
